src/gui/instrument/EnvelopeGraph.cpp

```diff
@@ -0,0 +1,157 @@
+/*
+ * EnvelopeGraph.cpp - Displays envelope graphs
+ *
+ * Copyright (c) 2004-2014 Tobias Doerffel <tobydox/at/users.sourceforge.net>
+ * Copyright (c) 2024-     Michael Gregorius
+ * 
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <QMouseEvent>
+#include <QPainter>
+
+#include "EnvelopeGraph.h"
+
+#include "EnvelopeAndLfoParameters.h"
+
+namespace lmms
+{
+
+namespace gui
+{
+
+const int TIME_UNIT_WIDTH = 40;
+
+
+EnvelopeGraph::EnvelopeGraph(QWidget* parent) :
+	QWidget(parent),
+	ModelView(nullptr, this),
+	m_params(nullptr)
+{
+	setFixedSize(m_envGraph.size());
+}
+
+void EnvelopeGraph::modelChanged()
+{
+	m_params = castModel<EnvelopeAndLfoParameters>();
+}
+
+void EnvelopeGraph::mousePressEvent(QMouseEvent* me)
+{
+	if(me->button() == Qt::LeftButton)
+	{
+		toggleAmountModel();
+	}
+}
+
+void EnvelopeGraph::paintEvent(QPaintEvent*)
+{
+	QPainter p(this);
+	p.setRenderHint(QPainter::Antialiasing);
+
+	// Draw the graph background
+	p.drawPixmap(rect(), m_envGraph);
+
+	const auto * params = castModel<EnvelopeAndLfoParameters>();
+	if (!params)
+	{
+		return;
+	}
+
+	const float amount = params->getAmountModel().value();
+	const float predelay = params->getPredelayModel().value();
+	const float attack = params->getAttackModel().value();
+	const float hold = params->getHoldModel().value();
+	const float decay = params->getDecayModel().value();
+	const float sustain = params->getSustainModel().value();
+	const float release = params->getReleaseModel().value();
+
+	const float gray_amount = 1.0f - fabsf(amount);
+
+	p.setPen(QPen(QColor(static_cast<int>(96 * gray_amount),
+				static_cast<int>(255 - 159 * gray_amount),
+				static_cast<int>(128 - 32 * gray_amount)),
+									2));
+
+	const QColor end_points_color(0x99, 0xAF, 0xFF);
+	const QColor end_points_bg_color(0, 0, 2);
```

Comment:
These variables are already renamed in another branch which is a follow-up PR to this one. The branch completely rewrites the rendering of the envelope graph so that it can render at any widget size. Not done here to prevent unnecessary merge conflicts and confusion.